net/proxymux/mux.go

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package proxymux splits a net.Listener in two, routing SOCKS5
+// connections to one and HTTP requests to the other.
+//
+// It allows for hosting both a SOCKS5 proxy and an HTTP proxy on the
+// same listener.
+package proxymux
+
+import (
+	"io"
+	"net"
+	"sync"
+)
+
+// Mux accepts connections on l and passes connections through to
+// either socksListener or httpListener, depending the first byte sent
+// by the client.
+func Mux(l net.Listener) (socksListener, httpListener net.Listener) {
+	sl := &listener{
+		addr:   l.Addr(),
+		c:      make(chan net.Conn),
+		closed: make(chan struct{}),
+	}
+	hl := &listener{
+		addr:   l.Addr(),
+		c:      make(chan net.Conn),
+		closed: make(chan struct{}),
+	}
+
+	go func() {
```

Comment:
make this a method on something so it shows up nicely in goroutine dumps? otherwise it's hard to keep track of tons of closure goroutines.